src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -31,83 +33,334 @@ constexpr float DEFAULT_QUEUE_PRIORITY = 1.0f;
 
-    const auto selected_gpu =
-        VulkanSettingsDecisionMaker::graphics_card(instance.instance(), surface, preferred_physical_device_index);
+    const bool is_swapchain_supported =
+        surface == nullptr || is_extension_supported(extensions, VK_KHR_SWAPCHAIN_EXTENSION_NAME);
 
-    if (!selected_gpu) {
-        throw std::runtime_error("Error: Could not find suitable graphics card!");
+    return DeviceInfo{
+        .name = properties.deviceName,
+        .physical_device = physical_device,
+        .type = properties.deviceType,
+        .total_device_local = total_device_local,
+        .presentation_supported = presentation_supported == VK_TRUE,
+        .swapchain_supported = is_swapchain_supported,
+        .features = features,
+        .extensions = extensions,
+    };
+}
+
+} // namespace
+
+VkPhysicalDevice Device::pick_best_physical_device(std::vector<DeviceInfo> &&physical_device_infos,
+                                                   const VkPhysicalDeviceFeatures &required_features,
+                                                   const std::vector<const char *> &required_extensions) {
+    if (physical_device_infos.empty()) {
+        throw std::runtime_error("Error: There are no physical devices available!");
+    }
+    std::sort(physical_device_infos.begin(), physical_device_infos.end(), [&](const auto &lhs, const auto &rhs) {
+        return compare_physical_devices(required_features, required_extensions, lhs, rhs);
+    });
+    if (!is_device_suitable(physical_device_infos.front(), required_features, required_extensions, true)) {
+        throw std::runtime_error("Error: Could not determine a suitable physical device!");
     }
+    return physical_device_infos.front().physical_device;
+}
+
+VkPhysicalDevice Device::pick_best_physical_device(const Instance &inst, const VkSurfaceKHR surface,
+                                                   const VkPhysicalDeviceFeatures &required_features,
+                                                   const std::vector<const char *> &required_extensions) {
+    // Put together all data that is required to compare the physical devices
+    const auto physical_devices = vk_tools::get_all_physical_devices(inst.instance());
+    std::vector<DeviceInfo> infos(physical_devices.size());
+    std::transform(physical_devices.begin(), physical_devices.end(), infos.begin(),
+                   [&](const VkPhysicalDevice physical_device) { return build_device_info(physical_device, surface); });
+    return pick_best_physical_device(std::move(infos), required_features, required_extensions);
+}
 
-    m_physical_device = *selected_gpu;
+Device::Device(const Instance &inst, const VkSurfaceKHR surface, const bool prefer_distinct_transfer_queue,
+               const VkPhysicalDevice physical_device, const VkPhysicalDeviceFeatures &required_features,
+               const std::vector<const char *> &required_extensions, const VkPhysicalDeviceFeatures &optional_features)
```

Comment:
I would now, in opposite as said before, not check for if the device is actually suitable. As i find it pointless that you pass **both** the device and what you want to check against to this constructor.
Either
1. hard code the required excentions and features into device.
2. do not check again in device and trust the developer that the device was suitable (maybe leave a warning in the docs)

Reply:
Nevermind. This is wrong.